core/shared/src/main/scala/zio/internal/ConcurrentWeakHashSet.scala

```diff
@@ -0,0 +1,579 @@
+
+    private val queue        = new ReferenceQueue[V]()
+    private val counter      = new AtomicInteger(0)
+    @volatile var references = new Array[Ref[V]](this.initialSize)
+
+    /**
+     * Get reference from this segment for given element and hash.
+     */
+    def getReference(element: V, hash: Int): Ref[V] = {
+      if (this.counter.get() == 0) return null
+      val localReferences = this.references // read volatile
+      val index           = this.getIndex(localReferences, hash)
+      val head            = localReferences(index)
+      this.findInChain(head, element, hash)
+    }
+
+    /**
+     * Get hash position in the references array.
+     */
+    private def getIndex(refs: Array[_], hash: Int): Int =
+      hash & (refs.length - 1)
+
+    /**
+     * Look for reference with given value in the chain.
+     */
+    private def findInChain(headReference: Ref[V], element: V, hash: Int): Ref[V] = {
+      var currentReference = headReference
+      while (currentReference != null) {
+        if (currentReference.hash == hash && currentReference.get() == element) {
+          return currentReference
+        }
+        currentReference = currentReference.nextRef
+      }
+      null
+    }
+
+    /**
+     * Perform update operation on matched reference for provided element in the
+     * set.
+     */
+    def update(hash: Int, element: V, update: ConcurrentWeakHashSet.UpdateOperation[V]): Boolean = {
+      val resize = update.options.contains(ConcurrentWeakHashSet.UpdateOptions.Resize)
+      if (update.options.contains(ConcurrentWeakHashSet.UpdateOptions.RestructureBefore)) {
+        this.restructureIfNecessary(resize)
+      }
+      val skipIfEmpty = update.options.contains(ConcurrentWeakHashSet.UpdateOptions.SkipIfEmpty)
+      if (skipIfEmpty && this.counter.get() == 0) {
+        return update.execute(null) != ConcurrentWeakHashSet.UpdateResults.None
+      }
+      this.lock()
+      try {
+        val index     = this.getIndex(this.references, hash)
+        val head      = this.references(index)
+        val storedRef = this.findInChain(head, element, hash)
+        update.execute(storedRef) match {
+          case ConcurrentWeakHashSet.UpdateResults.None =>
+            false
+          case ConcurrentWeakHashSet.UpdateResults.AddElement =>
+            val newRef = new Ref[V](hash, element, this.queue, head)
+            this.references(index) = newRef
+            this.counter.incrementAndGet()
+            true
+          case ConcurrentWeakHashSet.UpdateResults.RemoveElement =>
+            var previousRef = null.asInstanceOf[Ref[V]]
+            var currentRef  = storedRef
+            while (currentRef != null) {
+              if (currentRef.get() == element) {
+                this.counter.decrementAndGet()
+                if (previousRef == null) {
+                  this.references(index) = currentRef.nextRef // start chain with next ref
```

Comment:
I will do a more detailed review of this if we can find a way to avoid `ReentrantLock` (the logic may change somewhat so no need to review just yet).